modules/dnn/src/dnn.cpp

```diff
@@ -2374,10 +2337,57 @@ struct Net::Impl : public detail::NetImplBase
 #endif
     }
 
-    void initCUDABackend(const std::vector<LayerPin>& blobsToKeep_) {
+    void initCUDABackend(const std::vector<LayerPin>& blobsToKeep_)
+    {
         CV_Assert(haveCUDA());
+        CV_Assert(preferableBackend == DNN_BACKEND_CUDA);
 
 #ifdef HAVE_CUDA
+        if (cuda4dnn::getDeviceCount() <= 0)
+            CV_Error(Error::StsError, "No CUDA capable device found.");
+
+        if (cuda4dnn::getDevice() < 0)
+            CV_Error(Error::StsError, "No CUDA capable device selected.");
+
+        if (!cuda4dnn::isDeviceCompatible())
+            CV_Error(Error::GpuNotSupported, "OpenCV was not built to work with the selected device. Please check CUDA_ARCH_PTX or CUDA_ARCH_BIN in your build configuration.");
+
+        if (preferableTarget == DNN_TARGET_CUDA_FP16 && !cuda4dnn::doesDeviceSupportFP16())
+            CV_Error(Error::StsError, "The selected CUDA device does not support FP16 operations.");
+
+        if (!cudaInfo)
+        {
+            cuda4dnn::csl::CSLContext context;
+            context.stream = cuda4dnn::csl::Stream(true);
+            context.cublas_handle = cuda4dnn::csl::cublas::Handle(context.stream);
+            context.cudnn_handle = cuda4dnn::csl::cudnn::Handle(context.stream);
+
+            auto d2h_stream = cuda4dnn::csl::Stream(true); // stream for background D2H data transfers
+            cudaInfo = std::unique_ptr<CudaInfo_t>(new CudaInfo_t(std::move(context), std::move(d2h_stream)));
+            cuda4dnn::checkVersions();
```

Comment:
Massive facepalm. If a version mismatch occurs, exceptions will be thrown even before it reaches this line. This makes this version check mechanism completely useless.

Reply:
Could you move the check to proper place with another PR then?

Reply:
Yes, I am planning to do that. This time I am trying to test by making lots of containers with broken builds. So this will take some time.